### ISSUE TITLE
Minju: Separation-of-concerns, 3 weeks

### ISSUE THREAD

**Issue by minjupgeorge** [lab-brussels-1/home#374]
## Learning Objectives

<details>
<summary>Priorities: 🥚, 🐣, 🐥, 🐔 (click to learn more)</summary>
<br>

There is a lot to learn in this repository. If you can't master all the material
at once, that's expected! Anything you don't master now will always be waiting
for you to review when you need it. These 4 emoji's will help you prioritize
your study time and to measure your progress:

- 🥚: Understanding this material is required, it covers the base skills you'll
  need to move on. You do not need to finish all of them but should feel
  comfortable that you could with enough time.
- 🐣: You have started all of these exercises and feel you could complete them
  all if you just had more time. It may not be easy for you but with effort you
  can make it through.
- 🐥: You have studied the examples and started some exercises if you had time.
  You should have a big-picture understanding of these concepts/skills, but may
  not be confident completing the exercises.
- 🐔: These concepts or skills are not necessary but are related to this module.
  If you are finished with 🥚, 🐣 and 🐥 you can use the 🐔 exercises to push
  yourself without getting distracted from the module's main objectives.

---

</details>

- [ ] 🥚 **event-driven programming**: identify the concept in a JS program via listeners & handlers
- [ ]🥚 **entry points**: describe what an entry point is, there are 2 kinds in the programs for this module - initialization & interaction. identify them in a program
- 🥚 **function roles**: describe what function roles are and why they're important. they can identify a function's role given checklists for each role covered in this module:
  - [ ] **_listeners_**: functions that attach event listeners to the DOM
  - [ ] 🥚 **_handlers_**: entry point for user interactions
  - [ ] 🥚 **_utils_** _(utilities)_: pure functions to help do things with data
  - [ ] 🐣 **_components_**: render data into DOM elements to display for the user
  - [ ] 🐥 **_custom events_**: create custom events events in your components, passing useful data between components and handlers
- [ ] 🥚 **DOM access**: You read and write values from the DOM in an event handler
- [ ] 🥚 **es5 vs. es6**: You can demonstrate the change in developer-experience pre and post es6 by stepping through in the debugger and explaining differences in scoping \(global vs. script, block vs. local, modules\) between two programs with identical user experience but different implementations.
- [ ] 🥚 **Scope Hierarchy**: You is comfortable navigating different scopes in the browser's debugger to understand an existing application \(script, module, closure, local, block\)
- [ ] 🥚 **Code Splitting**: You can use ES Modules to split your code into multiple files & folders according to function role, data, listeners and initialization. They can use generated dependency diagrams and documentation to understand and navigate this folder structure.
- [ ] 🥚 **Dependency Graphs**: You can use a project's dependency graph to understand how it is organized and to navigate the source code.
- [ ] 🥚 **Development Strategies**: You can write development strategies that have all of the program's state defined at the beginning, and separate each user story into _interface_ and _interaction_ tasks.
- [ ] 🐣 **Naming Functions**: You can come up with clear and helpful names for the functions in your program. A good function name will take into account the function's role and the program's domain, like in the `/naming-variables` exercises from Debugging.
- [ ] 🐣 **DOM manipulation**: You can manipulate the DOM when implementing level-appropriate user interactions
- [ ] 🐣 **Isolating Components**: You can use a `test.html` file to render your components with different inputs
- [ ] 🐣 **Forms**: You can do basic handling of form data via `event.target.form`
- [ ] 🐣 **Handling events**: You can use the `event` argument to process user interactions, including bubbled events using `event.target`
- [ ] 🐣 **Passing Component Unit Tests**: You can write vanilla DOM component functions to pass provided unit tests
- [ ] 🐣 **refactoring**: refactor a single-script tutorial-style web page into multiple files using imports and exports
- [ ] 🐥 **reverse-engineering**: You can incrementally reverse-engineer a level-appropriate user interaction following these steps:
  - _init_
  - Listeners
  - Handlers
  - (possibly): Utils, Components, Custom Events
- [ ] 🐔 **From Spec**: given user stories, You can develop a site from scratch using a template repository.
- [ ] 🐔 **Writing Component Unit Tests**: You can write unit tests to validate your component functions using BDD syntax

**Comment by minjupgeorge** [comment]
Week 1

## Need help
First I took some time to understand what I have learned on Sunday and what I have to practice. I took so much of time for understanding DOM Manipulation

## What went well
I have gone through the examples. still I need some time to complete exercises. 
![image](https://user-images.githubusercontent.com/32439811/181601223-41df834b-b37c-4678-9d6e-58084bfb728f.png)


## Sunday Prep Work
First I wish to complete the exercises first.

**Comment by danielhalasz** [comment]
> For every html file we have init.js. do we need to write state.js , constants.js for every html file ?

no, these are created per project.. the point being that all parts of the app can access the same state/constants.

**Comment by minjupgeorge** [comment]
Week 1

## I Need Help With:

This week first two days I was struggling with my group about the group assignment. after several communication we have arrived to a conclusion what is the project about and how to be done. It would be better if we get a clear idea about the assignment on Sunday after the class. Hope that our doubts will be clarified during the Zoom meet today. 


## What went well?
1) I understood the concept of 'Separation of concern'. I was practicing the exercises under refactoring  and successfully completed flipflop, mouse over. still type casting is not working. am trying to make it run.
![image](https://user-images.githubusercontent.com/32439811/182832684-63b1528e-f1fb-466f-9bec-cb5aef61b4bd.png)

2) My group project is going well so far and we are about to complete the planning folder


## What went less well?
Still we have doubts regarding the assignment and we hope it will clear soon.


## Sunday Prep Work

**Comment by danielhalasz** [comment]
> Hope that our doubts will be clarified during the Zoom meet today.

please do not wait until a call next time and just ask in public or private on Slack

**Comment by minjupgeorge** [comment]
Week 3

## I Need Help With:

This week went fruitful that I was able to do some exercises still need to be more confident with DOM manipulation.

## What went well?

I am very happy to work with team Group2 for separation of concern. Its always great that you have companion to solve the errors. even-though we have separated few lines of code still its a great motivation for the rest of the project. 
This week self-study time was so precious that 8 hrs of work just we completed in 2hrs. Thanks a lot @danielhalasz  for the session as well as for making us clear what the separation of concern is.


## What went less well?

Our project just completed three user stories .Yet to modify CSS,HTML and two more user stories.


## Lessons Learned

How to separate core logic(input,logic and output )  in to different files.
Team Work is great. Motivation from the group will help us to solve the problems easily. 


## Sunday Prep Work
Yet to watch JS Event Loop.

**Comment by minjupgeorge** [comment]
## Me

### What was not clear, where did you get stuck?

At the beginning of separation of concern module,  concept was really difficult. Once started solving the exercises it became more familiar and now the concept is clear. still there are so many ways of DOM Manipulation. yet to practice mor 


### What was clear, what did you master?

Throughout the module, I became more and more familiar with separation of concerns concept  by examples and got confidence from the group. we felt so great that we were able to separate the files.

Now I am familiar with:

- `Refactoring'
- `Handling events`
- `reverse-engineering`

### Where can you help others moving forwards?

I am always try to make sure what is my team mate status of work. where he/she stuck up with. Zoom call meetings once in a day is a trigger call for my work as well group work. The best way of learning is to teach what we understood to our Team.

 ---

## The Course

### What can there be more of?

Exercises and projects need to be review. The concept of self-study time is great. 

### What can there be less of?

-
### What material were most helpful (from HYF or elsewhere)?

Honestly, I didn't even able  to complete  HYF study materials . There are lots of useful videos and websites  in each module repo. Needn't to go for other materials .

### Any suggestions for future classes?

Need to spend minimum three hours daily. Need to practice the exercises and need to prepared to the upcoming classes.

**Comment by danielhalasz** [comment]
> I am very happy to work with team Group2 for separation of concern. Its always great that you have companion to solve the errors. even-though we have separated few lines of code still its a great motivation for the rest of the project.
> This week self-study time was so precious that 8 hrs of work just we completed in 2hrs. Thanks a lot @danielhalasz for the session as well as for making us clear what the separation of concern is.

💟

**Comment by danielhalasz** [comment]
> Our project just completed three user stories .Yet to modify CSS,HTML and two more user stories.

don't worry about not finishing it. being more comfortable with the process and the concept itself is the key

**Comment by danielhalasz** [comment]
> Exercises and projects need to be review.

absolutely agree, wish we could do more of that.. you also always tag coaches to review your work, once it is ready for it or to write them a message on Slack
